src/main/java/org/apache/commons/text/cases/CamelCase.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.commons.text.cases;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.commons.lang3.CharUtils;
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ * Case implementation that parses and formats strings of the form 'myCamelCase'
+ * <p>
+ * This case separates tokens on uppercase ASCII alpha characters. Each token begins with an
```

Comment:
If you're talking about the logic of dividing tokens, my understanding is that the uppercase/lowercase distinction, which is inherently necessary for proper camelCase, is a quirk of the Latin languages. If a similar distinction is present in other languages that require unicode characters then I suppose this class could be broken into two classes - a base class that accepts character ranges and then concrete classes that provide them. This would allow users to implement a Case that works like camelCase but divides the tokens based on some other character range. But as far as camelCase specifically, I think it makes sense to stick to the accepted definition.

To my surprise, [many modern languages support unicode identifiers](https://rosettacode.org/wiki/Unicode_variable_names). Which means we'd definitely want to continue to handle unicode for Case implementations that dont depend on upper/lower case characters.

Reply:
ASCII != Latin. There are many upper case letters in French, Icelandic Turkish, etc. not handled by ASCII. I think the Unicod spec normally identifies each character as upper case, lower case, and something else for non-cased characters.

Definitely Greek and I think Cyrillic also have case. I'm sure there are others.

Also, ranges might not do what you expect.

Reply:
That's good to know. Im definitely not a unicode expert, so I'll defer to you on all of those points. Seems like using the SDK methods: `Character.is<Upper|Lower>Case` and` Character.to<Upper|Lower>Case` would work? Docs say they can handle unicode cases. I believe that will work and I can't think of any disadvantages or ill effects it would have on the ascii use case.

Although I can manually come up with test cases, It would be nice if there were some real world code that used non-ascii identifiers, but  can't seem to find any (other than examples of how one _might_ use them).

Reply:
Regarding handling conversion with the Turkish locales: https://garygregory.wordpress.com/2015/11/03/java-lowercase-conversion-turkey/

Reply:
Per the docs, I believe the Character.toXXXCase methods don't use locale. So the question then is do we want to try and support Locale?

Reply:
Yes they do. You pass the target locale to the [one arg method](https://docs.oracle.com/javase/8/docs/api/java/lang/String.html#toUpperCase-java.util.Locale-)  otherwise the default locale is used when you call the no arg method.

Reply:
That's the String class method. The Character class methods don't accept a locale. Looking at the source I don't think they delegate to String anywhere.

Reply:
This is tricky. It's been a while since I needed this but I seem to recall that there is a canonical Unicode case mapping in Unicode Data somewhere that would fit here. I wouldn't use a Locale for this, implicitly or explicitly. I think what we need is whether the character has the Lu property, the Lc property, or neither. 

See https://en.wikipedia.org/wiki/Unicode_character_property

Character.isUpperCase and Character.isLowerCase should be essentially this.

Reply:
> That's the String class method. The Character class methods don't accept a locale. Looking at the source I don't think they delegate to String anywhere. The docs advise using String specifically to support Locale. But if we decided not to support Locale the the Character methods might work?

Ah, yes of course. Thank you for pointing that out :-)

Reply:
@elharo I like that approach. Shouldn't be too big of a change either.

Reply:
So, in making this change, I ran into another quirk. Unicode has 3 cases (upper, lower, and title). In my updated implementation, I was only handling upper and lower. i.e. in order to convert the token `Xabc` into a Pascal string, the `X` character _must_ translate to (or already be) an uppercase character. However, one might argue that having a titlecase character is also sufficient, since the purpose of the Pascal case is to uppercase only the first character, which is also the purpose of unicode titlecase.

So this string `dzabc` (note that "dz" is the single digraph character) could be represented as either:

- DZabc (dz to uppercase)
- Dzabc (dz to titlecase)

For that specific character we have two possible options, however there are also lowercase characters whose uppercase character is of the **titlecase** category. Which means anywhere we use the `Character.isUpperCase` check, it will fail, despite the fact that kinda works.

I see a couple questions / decisions...

- Should we allow titlecase characters to serve as uppercase?
- Should we _prefer_ unicode titlecase over uppercase when both exist?
- Should the user be able to choose?

(In all cases, if a character cannot be mapped to lowercase, then logic that requires it will throw an exception).

Reply:
The on ly thing I'm sure of is that the user should not be able to choose. Make it deterministic. I think but am not sure you should stick to uppercase when both exist. 

Since this is primarily for programming I think we need to understand and treat digraph characters as single characters, even if some folks might read them otherwise. That is, the digraph DZ is treated differently than the two letters D and Z. But again, I am not sure.

Reply:
I updated the Camel/Pascal case classes significantly to rely on the [unicode case cateogry](https://www.unicode.org/L2/L1999/UnicodeData.html#General%20Category). 

Any character that needs to be converted to a upper / lower case must convert precisely to that case (as implemented by `Character.isUpperCase` and `Character.isLowerCase`). This is worth noting because a few unicode characters have upper case mappings point to a title case character which has no clear lower case mapping, and thus cannot be deterministically converted back to lower case (without some sort of user configuration).

Updated test cases accordingly to include digraph character and title case character tests.